### DISCUSSION TITLE
Roadmap for pypdf==4.0.0

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#2041] [General]
Hey :-)

I want to release `pypdf==4.0.0` in 2024 (but not earlier). This is a check-list for me. Please let me know if I should add anything.

I want to do the following with `pypdf==4.0.0` (but not earlier):

* **Cryptography**: Set the default algorithm to be something we consider secure, e.g. `AES-256` (see `encryption-decryption.md`). That is a breaking change as it means the `cryptography` package will be required.
* :heavy_check_mark:  https://github.com/py-pdf/pypdf/pull/2367 - **Deprecations**: Get rid of a lot of deprecated code :tada: 
    * Search for `deprecate_with_replacement(removed_in="4.0.0")` : All of those function calls need to be replaced by `deprecation_with_replacement`.
    * Search for `deprecate_no_replacement(removed_in="4.0.0")`: All of those function calls need to be replaced with `deprecation_no_replacement`
    * Search for `deprecation_with_replacement(removed_in="3.0.0")`: All of those functions / modules can be completely removed. Remove the default value of the function itself.
    * Search for `deprecation_no_replacement(removed_in="3.0.0")`: All of those functions / modules can be completely removed. Remove the default value of the function itself.
* :heavy_check_mark:  https://github.com/py-pdf/pypdf/pull/2376 - Clarify https://github.com/py-pdf/pypdf/discussions/2010
* :heavy_check_mark:  https://github.com/py-pdf/pypdf/pull/2369 -  https://github.com/py-pdf/pypdf/discussions/2005 - we don't want major version changes that often. The PIL issues and lacking crytpography support are a good reason for me to drop 3.6 support. 
* Go through all https://github.com/py-pdf/pypdf/labels/breaking-change issues / [discussions](https://github.com/py-pdf/pypdf/discussions?discussions_q=is%3Aopen+label%3Abreaking-change)

**Comment by MartinThoma** [comment]
We could **update the Sphinx theme** as a signal that we're getting more modern. My requirements would be:

* Makes use of an free software license (MIT, BSD, Apache 2, ...)
* [Autosphinx generated stuff](https://pypdf.readthedocs.io/en/latest/modules/Transformation.html) must look good
* [Code blocks](https://pypdf.readthedocs.io/en/latest/user/add-watermark.html) have to be still syntax highlighted / easy to read
* [Search results](https://pypdf.readthedocs.io/en/3.14.0/search.html?q=annotations&check_keywords=yes&area=default) must look good
* A light-ish mode must be the default, a dark mode can exist
* It should work with Desktop and mobile devices
* If you have indications that something is better for vision-impaired people, let me know (e.g. easier for screen readers)

I know that "must look good" is very subjective. I'm open to suggestions and community upvotes :+1:  / downovtes :-1:  certainly influence me, but I will make the final decision. 

What I found:

* https://sphinxawesome.xyz/
* https://sphinxthemes.com/themes/guzzle
* https://sphinxthemes.com/themes/furo - I like the search + it has a dark mode

**Reply by stefan6419846** [reply]
In my personal opinion, the default RTD template we currently use is completely fine for most of my use cases, so this does not really have any real priority for me.

**Comment by MartinThoma** [comment]
@pubpub-zz @stefan6419846 On first or second of January 2024 I'll add bigger PRs to do the deprecations. I will go through the open PRs beforehand and try to merge as many as possible.

This is just a heads-up. Doing the deprecations will likely give several merge conflicts. I don't want to annoy you / cause unnecessary work for you.

I'm looking forward to those cleanup-changes xD

**Reply by pubpub-zz** [reply]
I will try to #2197 around Christmas : It should be better to have it in 4.0.0 as some changes should occur in the interface

**Reply by stefan6419846** [reply]
I have been a bit confused at first why I am mentioned here, but I assume that this is due to being one of the continuous contributors here as well, although mostly related to smaller stuff ;)

For me, this is completely fine. (Most of my issues are work-related and thus actual contribution activity tends to be less on holidays anyway.)
